### PR TITLE
Let configuration cache support composite builds with builds included in any order

### LINKED ISSUE #20945: Composite builds fail when reusing configuration cache
### Expected Behavior

Composite builds should work with `--configuration-cache`, since it's not documented that they don't and they appear to work in at least some cases.

### Current Behavior

I have a repro that shows included builds not being found when the configuration cache is reused. First build (stores the cache entry):

```
[0] ~/repro/CompositeBuild # ./gradlew clean build --configuration-cache
[...]
BUILD SUCCESSFUL in 5s
4 actionable tasks: 4 executed
Configuration cache entry stored.
```

Second build (reuses the cached entry):

```
[0] ~/repro/CompositeBuild # ./gradlew clean build --configuration-cache
Configuration cache is an incubating feature.
Reusing configuration cache.

FAILURE: Build failed with an exception.

* What went wrong:
Could not find build 'Util'
```

### Context

I work with a few large composite builds and I want to take advantage of the configuration cache.

### Steps to Reproduce

Extract and run the reproducer, [repro.tar.gz](https://github.com/gradle/gradle/files/8847962/repro.tar.gz):

```
$ tar -zxvf repro.tar.gz
$ cd CompositeBuild
$ ./gradlew clean build --configuration-cache
$ ./gradlew clean build --configuration-cache
```

### Your Environment

I've reproduced this on both Linux and macOS, on both case-sensitive and case-insensitive volumes, with multiple versions of Java, using Gradle 7.4.2 as well as the latest Gradle nightly.

### PR DESCRIPTION
**Pull request by bamboo**
By lazily locating tasks from another build after all builds have been loaded.

Fixes #20945

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test APT

**Comment by bamboo** [comment]
@bot-gradle test and merge

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheCompositeBuildsIntegrationTest.groovy [hunk 1]
```diff
@@ -270,6 +271,60 @@ class ConfigurationCacheCompositeBuildsIntegrationTest extends AbstractConfigura
...
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheRun 'clean', 'build'
+
+        then:
+        configurationCache.assertStateLoaded()
+
+        and:
+        result.assertTaskOrder(':lib:build', ':util:build', ':build')
```
Comment (mlopatkin) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheCompositeBuildsIntegrationTest.groovy:322]: Looks like having a dependency [isn't enough](https://ge.gradle.org/scans/tests?search.timeZoneId=Europe/Berlin&tests.container=org.gradle.configurationcache.ConfigurationCacheCompositeBuildsIntegrationTest&tests.test=composite%20build%20with%20dependency%20substitution%20can%20include%20builds%20in%20any%20order%20%5Border:%20%5Blib%2C%20util%5D%2C%20%230%5D) to ensure the declared order of `:build` tasks. I suppose, the exported configuration of `lib` is built by `assemble` or some other lower-level task.

Comment (bamboo) [reply subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheCompositeBuildsIntegrationTest.groovy:322]: Indeed. `compileJava` is the right one.

**Pull request merged**